// ng2-ckeditor/bundles/ng2-ckeditor.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('ng2-ckeditor', ['exports', '@angular/core', '@angular/common', '@angular/forms'], factory) :
    (global = global || self, factory(global['ng2-ckeditor'] = {}, global.ng.core, global.ng.common, global.ng.forms));
}(this, function (exports, core, common, forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    /**
     * CKGroup component
     * Usage :
     *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
     *      <ckbutton [name]="'SaveButton'" [command]="'saveCommand'" (click)="save($event)"
     *                [icon]="'/save.png'" [toolbar]="'customGroup,1'" [label]="'Save'">
     *      </ckbutton>
     *   </ckeditor>
     */
    var CKButtonDirective = /** @class */ (function () {
        function CKButtonDirective() {
            this.click = new core.EventEmitter();
        }
        CKButtonDirective.prototype.initialize = function (editor) {
            var _this = this;
            editor.instance.addCommand(this.command, {
                exec: function (evt) {
                    _this.click.emit(evt);
                },
            });
            editor.instance.ui.addButton(this.name, {
                label: this.label,
                command: this.command,
                toolbar: this.toolbar,
                icon: this.icon,
            });
        };
        CKButtonDirective.prototype.ngOnInit = function () {
            if (!this.name)
                throw new Error('Attribute "name" is required on <ckbutton>');
            if (!this.command)
                throw new Error('Attribute "command" is required on <ckbutton>');
        };
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKButtonDirective.prototype, "click", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "label", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "command", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "toolbar", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "name", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "icon", void 0);
        CKButtonDirective = __decorate([
            core.Directive({
                selector: 'ckbutton',
            })
        ], CKButtonDirective);
        return CKButtonDirective;
    }());

    /**
     * CKGroup component
     * Usage :
     *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
     *      <ckgroup [name]="'exampleGroup2'" [previous]="'1'" [subgroupOf]="'exampleGroup1'">
     *          .
     *          .
     *      </ckgroup>
     *   </ckeditor>
     */
    var CKGroupDirective = /** @class */ (function () {
        function CKGroupDirective() {
        }
        CKGroupDirective.prototype.ngAfterContentInit = function () {
            var _this = this;
            // Reconfigure each button's toolbar property within ckgroup to hold its parent's name
            this.toolbarButtons.forEach(function (button) { return (button.toolbar = _this.name); });
        };
        CKGroupDirective.prototype.initialize = function (editor) {
            editor.instance.ui.addToolbarGroup(this.name, this.previous, this.subgroupOf);
            // Initialize each button within ckgroup
            this.toolbarButtons.forEach(function (button) {
                button.initialize(editor);
            });
        };
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKGroupDirective.prototype, "name", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], CKGroupDirective.prototype, "previous", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKGroupDirective.prototype, "subgroupOf", void 0);
        __decorate([
            core.ContentChildren(CKButtonDirective),
            __metadata("design:type", core.QueryList)
        ], CKGroupDirective.prototype, "toolbarButtons", void 0);
        CKGroupDirective = __decorate([
            core.Directive({
                selector: 'ckgroup',
            })
        ], CKGroupDirective);
        return CKGroupDirective;
    }());

    /**
     * CKEditor component
     * Usage :
     *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500"></ckeditor>
     */
    var CKEditorComponent = /** @class */ (function () {
        /**
         * Constructor
         */
        function CKEditorComponent(zone) {
            this.zone = zone;
            this.change = new core.EventEmitter();
            this.editorChange = new core.EventEmitter();
            this.ready = new core.EventEmitter();
            this.blur = new core.EventEmitter();
            this.focus = new core.EventEmitter();
            this.contentDom = new core.EventEmitter();
            this.fileUploadRequest = new core.EventEmitter();
            this.fileUploadResponse = new core.EventEmitter();
            this.paste = new core.EventEmitter();
            this.drop = new core.EventEmitter();
            this._value = '';
        }
        CKEditorComponent_1 = CKEditorComponent;
        Object.defineProperty(CKEditorComponent.prototype, "value", {
            get: function () {
                return this._value;
            },
            set: function (v) {
                if (v !== this._value) {
                    this._value = v;
                    this.onChange(v);
                }
            },
            enumerable: true,
            configurable: true
        });
        CKEditorComponent.prototype.ngOnChanges = function (changes) {
            if (changes.readonly && this.instance) {
                this.instance.setReadOnly(changes.readonly.currentValue);
            }
        };
        /**
         * On component destroy
         */
        CKEditorComponent.prototype.ngOnDestroy = function () {
            var _this = this;
            if (this.instance) {
                setTimeout(function () {
                    if (_this.instance) {
                        _this.instance.removeAllListeners();
                        CKEDITOR.instances[_this.instance.name].destroy();
                        _this.instance.destroy();
                        _this.instance = null;
                    }
                });
            }
        };
        /**
         * On component view init
         */
        CKEditorComponent.prototype.ngAfterViewInit = function () {
            this.ckeditorInit(this.config || {});
        };
        /**
         * On component view checked
         */
        CKEditorComponent.prototype.ngAfterViewChecked = function () {
            this.ckeditorInit(this.config || {});
        };
        /**
         * Value update process
         */
        CKEditorComponent.prototype.updateValue = function (value) {
            var _this = this;
            this.zone.run(function () {
                _this.value = value;
                _this.onChange(value);
                _this.onTouched();
                _this.change.emit(value);
            });
        };
        /**
         * CKEditor init
         */
        CKEditorComponent.prototype.ckeditorInit = function (config) {
            var _this = this;
            if (typeof CKEDITOR === 'undefined') {
                console.warn('CKEditor 4.x is missing (http://ckeditor.com/)');
            }
            else {
                // Check textarea exists
                if (this.instance || !this.documentContains(this.host.nativeElement)) {
                    return;
                }
                if (this.readonly) {
                    config.readOnly = this.readonly;
                }
                // CKEditor replace textarea
                this.instance = CKEDITOR.replace(this.host.nativeElement, config);
                // Set initial value
                this.instance.setData(this.value);
                // listen for instanceReady event
                this.instance.on('instanceReady', function (evt) {
                    // if value has changed while instance loading
                    // update instance with current component value
                    if (_this.instance.getData() !== _this.value) {
                        _this.instance.setData(_this.value);
                    }
                    // send the evt to the EventEmitter
                    _this.ready.emit(evt);
                });
                // CKEditor change event
                this.instance.on('change', function (evt) {
                    _this.onTouched();
                    var value = _this.instance.getData();
                    if (_this.value !== value) {
                        // Debounce update
                        if (_this.debounce) {
                            if (_this.debounceTimeout)
                                clearTimeout(_this.debounceTimeout);
                            _this.debounceTimeout = setTimeout(function () {
                                _this.updateValue(value);
                                _this.debounceTimeout = null;
                            }, parseInt(_this.debounce));
                            // Live update
                        }
                        else {
                            _this.updateValue(value);
                        }
                    }
                    // Original ckeditor event dispatch
                    _this.editorChange.emit(evt);
                });
                // CKEditor blur event
                this.instance.on('blur', function (evt) {
                    _this.blur.emit(evt);
                });
                // CKEditor focus event
                this.instance.on('focus', function (evt) {
                    _this.focus.emit(evt);
                });
                // CKEditor contentDom event
                this.instance.on('contentDom', function (evt) {
                    _this.contentDom.emit(evt);
                });
                // CKEditor fileUploadRequest event
                this.instance.on('fileUploadRequest', function (evt) {
                    _this.fileUploadRequest.emit(evt);
                });
                // CKEditor fileUploadResponse event
                this.instance.on('fileUploadResponse', function (evt) {
                    _this.fileUploadResponse.emit(evt);
                });
                // CKEditor paste event
                this.instance.on('paste', function (evt) {
                    _this.paste.emit(evt);
                });
                // CKEditor drop event
                this.instance.on('drop', function (evt) {
                    _this.drop.emit(evt);
                });
                // Add Toolbar Groups to Editor. This will also add Buttons within groups.
                this.toolbarGroups.forEach(function (group) {
                    group.initialize(_this);
                });
                // Add Toolbar Buttons to Editor.
                this.toolbarButtons.forEach(function (button) {
                    button.initialize(_this);
                });
            }
        };
        /**
         * Implements ControlValueAccessor
         */
        CKEditorComponent.prototype.writeValue = function (value) {
            this._value = value;
            if (this.instance)
                this.instance.setData(value);
        };
        CKEditorComponent.prototype.onChange = function (_) { };
        CKEditorComponent.prototype.onTouched = function () { };
        CKEditorComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        CKEditorComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        CKEditorComponent.prototype.documentContains = function (node) {
            return document.contains ? document.contains(node) : document.body.contains(node);
        };
        var CKEditorComponent_1;
        CKEditorComponent.ctorParameters = function () { return [
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "config", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], CKEditorComponent.prototype, "readonly", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKEditorComponent.prototype, "debounce", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "change", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "editorChange", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "ready", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "blur", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "focus", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "contentDom", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "fileUploadRequest", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "fileUploadResponse", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "paste", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "drop", void 0);
        __decorate([
            core.ViewChild('host', { static: false }),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "host", void 0);
        __decorate([
            core.ContentChildren(CKButtonDirective),
            __metadata("design:type", core.QueryList)
        ], CKEditorComponent.prototype, "toolbarButtons", void 0);
        __decorate([
            core.ContentChildren(CKGroupDirective),
            __metadata("design:type", core.QueryList)
        ], CKEditorComponent.prototype, "toolbarGroups", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Object),
            __metadata("design:paramtypes", [Object])
        ], CKEditorComponent.prototype, "value", null);
        CKEditorComponent = CKEditorComponent_1 = __decorate([
            core.Component({
                selector: 'ckeditor',
                providers: [
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core.forwardRef(function () { return CKEditorComponent_1; }),
                        multi: true,
                    },
                ],
                template: "\n    <textarea #host></textarea>\n  "
            }),
            __metadata("design:paramtypes", [core.NgZone])
        ], CKEditorComponent);
        return CKEditorComponent;
    }());

    /**
     * CKEditorModule
     */
    var CKEditorModule = /** @class */ (function () {
        function CKEditorModule() {
        }
        CKEditorModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule],
                declarations: [CKEditorComponent, CKButtonDirective, CKGroupDirective],
                exports: [CKEditorComponent, CKButtonDirective, CKGroupDirective],
            })
        ], CKEditorModule);
        return CKEditorModule;
    }());

    exports.CKEditorModule = CKEditorModule;
    exports.ɵa = CKEditorComponent;
    exports.ɵb = CKButtonDirective;
    exports.ɵc = CKGroupDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
